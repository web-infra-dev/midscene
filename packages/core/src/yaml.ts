import type { PlanningActionParamScroll } from './types';

export interface LocateOption {
  prompt?: string;
  deepThink?: boolean; // only available in vl model
}

export interface DetailedLocateParam extends LocateOption {
  prompt: string;
}

export interface scrollParam {
  direction: 'down' | 'up' | 'right' | 'left';
  scrollType: 'once' | 'untilBottom' | 'untilTop' | 'untilRight' | 'untilLeft';
  distance?: null | number; // distance in px
}

export interface MidsceneYamlScript {
  target: MidsceneYamlScriptEnv;
  tasks: MidsceneYamlTask[];
}

export interface MidsceneYamlTask {
  name: string;
  flow: MidsceneYamlFlowItem[];
  continueOnError?: boolean;
}

export interface MidsceneYamlScriptEnv {
  output?: string;
  aiActionContext?: string;

  // for web only
  serve?: string;
  url: string;

  // puppeteer only
  userAgent?: string;
  viewportWidth?: number;
  viewportHeight?: number;
  viewportScale?: number;
  waitForNetworkIdle?: {
    timeout?: number; // ms, 30000 for default, set to 0 to disable
    continueOnNetworkIdleError?: boolean; // should continue if failed to wait for network idle, true for default
  };
  cookie?: string;
  forceSameTabNavigation?: boolean; // if track the newly opened tab, true for default in yaml script

  // bridge mode config
  bridgeMode?: false | 'newTabWithUrl' | 'currentTab';
  closeNewTabsAfterDisconnect?: boolean;
}

export interface MidsceneYamlFlowItemAIAction {
  ai?: string; // this is the shortcut for aiAction
  aiAction?: string;
  aiActionProgressTips?: string[];
}

export interface MidsceneYamlFlowItemAIAssert {
  aiAssert: string;
}

export interface MidsceneYamlFlowItemAIQuery {
  aiQuery: string;
  name?: string;
}

export interface MidsceneYamlFlowItemAIWaitFor {
  aiWaitFor: string;
  timeout?: number;
}

export interface MidsceneYamlFlowItemAITap extends LocateOption {
  aiTap: string;
}

export interface MidsceneYamlFlowItemAIHover extends LocateOption {
  aiHover: string;
}

export interface MidsceneYamlFlowItemAIInput extends LocateOption {
  aiInput: string; // value to input
  locate: string; // where to input
}

export interface MidsceneYamlFlowItemAIKeyboardPress extends LocateOption {
  aiKeyboardPress: string;
  locate?: string; // where to press, optional
}

export interface MidsceneYamlFlowItemAIScroll
  extends LocateOption,
    PlanningActionParamScroll {
  aiScroll: null;
  locate?: string; // which area to scroll, optional
}

export interface MidsceneYamlFlowItemSleep {
  sleep: number;
}

export type MidsceneYamlFlowItem =
  | MidsceneYamlFlowItemAIAction
  | MidsceneYamlFlowItemAIAssert
  | MidsceneYamlFlowItemAIQuery
  | MidsceneYamlFlowItemAIWaitFor
  | MidsceneYamlFlowItemAITap
  | MidsceneYamlFlowItemAIHover
  | MidsceneYamlFlowItemAIInput
  | MidsceneYamlFlowItemAIKeyboardPress
  | MidsceneYamlFlowItemAIScroll
  | MidsceneYamlFlowItemSleep;

export interface FreeFn {
  name: string;
  fn: () => void;
}

export interface ScriptPlayerTaskStatus extends MidsceneYamlTask {
  status: ScriptPlayerStatusValue;
  currentStep?: number;
  totalSteps: number;
  error?: Error;
}

export type ScriptPlayerStatusValue = 'init' | 'running' | 'done' | 'error';
