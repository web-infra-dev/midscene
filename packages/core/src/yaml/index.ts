export * from './player';
export * from './builder';
export * from './utils';
