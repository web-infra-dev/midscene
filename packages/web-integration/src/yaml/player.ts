import assert from 'node:assert';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { paramStr, typeStr } from '@/common/ui-utils';

import type { PageAgent } from '@/common/agent';
import type {
  FreeFn,
  MidsceneYamlFlowItemAIAction,
  MidsceneYamlFlowItemAIAssert,
  MidsceneYamlFlowItemAIQuery,
  MidsceneYamlFlowItemAIWaitFor,
  MidsceneYamlFlowItemSleep,
  MidsceneYamlScript,
  MidsceneYamlScriptEnv,
  ScriptPlayerStatusValue,
  ScriptPlayerTaskStatus,
} from '@midscene/core';

export class ScriptPlayer {
  public currentTaskIndex?: number;
  public taskStatusList: ScriptPlayerTaskStatus[] = [];
  public status: ScriptPlayerStatusValue = 'init';
  public reportFile?: string | null;
  public result: Record<string, any>;
  private unnamedResultIndex = 0;
  public output?: string | null;
  public errorInSetup?: Error;
  private pageAgent: PageAgent | null = null;
  constructor(
    private script: MidsceneYamlScript,
    private setupAgent: (target: MidsceneYamlScriptEnv) => Promise<{
      agent: PageAgent;
      freeFn: FreeFn[];
    }>,
    public onTaskStatusChange?: (taskStatus: ScriptPlayerTaskStatus) => void,
  ) {
    this.result = {};
    this.output = script.target.output;
    this.taskStatusList = (script.tasks || []).map((task, taskIndex) => ({
      ...task,
      index: taskIndex,
      status: 'init',
      totalSteps: task.flow?.length || 0,
    }));
  }

  private setPlayerStatus(status: ScriptPlayerStatusValue, error?: Error) {
    this.status = status;
    this.errorInSetup = error;
  }

  private notifyCurrentTaskStatusChange(taskIndex?: number) {
    const taskIndexToNotify =
      typeof taskIndex === 'number' ? taskIndex : this.currentTaskIndex;

    if (typeof taskIndexToNotify !== 'number') {
      return;
    }

    const taskStatus = this.taskStatusList[taskIndexToNotify];
    if (this.onTaskStatusChange) {
      this.onTaskStatusChange(taskStatus);
    }
  }

  private async setTaskStatus(
    index: number,
    statusValue: ScriptPlayerStatusValue,
    error?: Error,
  ) {
    this.taskStatusList[index].status = statusValue;
    if (error) {
      this.taskStatusList[index].error = error;
    }

    this.notifyCurrentTaskStatusChange(index);
  }

  private setTaskIndex(taskIndex: number) {
    this.currentTaskIndex = taskIndex;
  }

  private flushResult() {
    if (Object.keys(this.result).length && this.output) {
      const output = join(process.cwd(), this.output);
      const outputDir = dirname(output);
      if (!existsSync(outputDir)) {
        mkdirSync(outputDir, { recursive: true });
      }
      writeFileSync(output, JSON.stringify(this.result, undefined, 2));
    }
  }

  async playTask(taskStatus: ScriptPlayerTaskStatus, agent: PageAgent) {
    const { flow } = taskStatus;
    assert(flow, 'missing flow in task');
    const notifyTaskStatusChange =
      this.notifyCurrentTaskStatusChange.bind(this);

    for (const flowItemIndex in flow) {
      const currentStep = Number.parseInt(flowItemIndex, 10);
      taskStatus.currentStep = currentStep;
      const flowItem = flow[flowItemIndex];
      if (
        (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
        (flowItem as MidsceneYamlFlowItemAIAction).ai
      ) {
        const actionTask = flowItem as MidsceneYamlFlowItemAIAction;
        const prompt = actionTask.aiAction || actionTask.ai;
        assert(prompt, 'missing prompt for ai (aiAction)');
        assert(
          typeof prompt === 'string',
          'prompt for aiAction must be a string',
        );
        await agent.aiAction(prompt, {
          onTaskStart(task) {
            const tip = `${typeStr(task)} - ${paramStr(task)}`;
            const actionItem = flowItem as MidsceneYamlFlowItemAIAction;
            actionItem.aiActionProgressTips =
              actionItem.aiActionProgressTips || [];
            actionItem.aiActionProgressTips.push(tip);

            notifyTaskStatusChange();
          },
        });
      } else if ((flowItem as MidsceneYamlFlowItemAIAssert).aiAssert) {
        const assertTask = flowItem as MidsceneYamlFlowItemAIAssert;
        const prompt = assertTask.aiAssert;
        assert(prompt, 'missing prompt for aiAssert');
        assert(
          typeof prompt === 'string',
          'prompt for aiAssert must be a string',
        );
        await agent.aiAssert(prompt);
      } else if ((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery) {
        const queryTask = flowItem as MidsceneYamlFlowItemAIQuery;
        const prompt = queryTask.aiQuery;
        assert(prompt, 'missing prompt for aiQuery');
        assert(
          typeof prompt === 'string',
          'prompt for aiQuery must be a string',
        );
        const queryResult = await agent.aiQuery(prompt);
        const resultKey = queryTask.name || this.unnamedResultIndex++;
        if (this.result[resultKey]) {
          console.warn(
            `result key ${resultKey} already exists, will overwrite`,
          );
        }

        this.result[resultKey] = queryResult;
        this.flushResult();
      } else if ((flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor) {
        const waitForTask = flowItem as MidsceneYamlFlowItemAIWaitFor;
        const prompt = waitForTask.aiWaitFor;
        assert(prompt, 'missing prompt for aiWaitFor');
        assert(
          typeof prompt === 'string',
          'prompt for aiWaitFor must be a string',
        );
        const timeout = waitForTask.timeout;
        await agent.aiWaitFor(prompt, { timeoutMs: timeout });
      } else if ((flowItem as MidsceneYamlFlowItemSleep).sleep) {
        const sleepTask = flowItem as MidsceneYamlFlowItemSleep;
        const ms = sleepTask.sleep;
        assert(
          ms && ms > 0,
          `ms for sleep must be greater than 0, but got ${ms}`,
        );
        await new Promise((resolve) => setTimeout(resolve, ms));
      } else {
        throw new Error(`unknown flowItem: ${JSON.stringify(flowItem)}`);
      }
    }
    this.reportFile = agent.reportFile;
  }

  async run() {
    const { target, tasks } = this.script;
    this.setPlayerStatus('running');

    let agent: PageAgent | null = null;
    let freeFn: FreeFn[] = [];
    try {
      const { agent: newAgent, freeFn: newFreeFn } =
        await this.setupAgent(target);
      agent = newAgent;
      freeFn = newFreeFn;
    } catch (e) {
      this.setPlayerStatus('error', e as Error);
      return;
    }
    this.pageAgent = agent;

    let taskIndex = 0;
    this.setPlayerStatus('running');
    let errorFlag = false;
    while (taskIndex < tasks.length) {
      this.setTaskStatus(taskIndex, 'running' as any);

      try {
        this.setTaskIndex(taskIndex);
        await this.playTask(this.taskStatusList[taskIndex], this.pageAgent);
      } catch (e) {
        this.setTaskStatus(taskIndex, 'error' as any, e as Error);
        this.setPlayerStatus('error');
        errorFlag = true;

        this.reportFile = agent.reportFile;
        taskIndex++;
        continue;
      }
      this.reportFile = agent.reportFile;
      this.setTaskStatus(taskIndex, 'done' as any);
      taskIndex++;
    }

    if (!errorFlag) {
      this.setPlayerStatus('done');
    }

    // free the resources
    freeFn.forEach((fn) => {
      try {
        fn.fn();
      } catch (e) {}
    });
  }
}
