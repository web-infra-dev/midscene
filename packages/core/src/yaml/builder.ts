import yaml from 'js-yaml';
import type {
  MidsceneYamlScript,
  MidsceneYamlScriptWebEnv,
  MidsceneYamlTask,
} from '../types';

export function buildYaml(
  env: MidsceneYamlScriptWebEnv,
  tasks: MidsceneYamlTask[],
) {
  const result: MidsceneYamlScript = {
    target: env,
    tasks,
  };

  return yaml.dump(result, {
    indent: 2,
  });
}
