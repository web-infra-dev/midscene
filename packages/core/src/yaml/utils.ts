import type {
  DetailedLocateParam,
  LocateOption,
  MidsceneYamlScript,
  TUserPrompt,
} from '@/index';
import { getDebug } from '@midscene/shared/logger';
import { assert } from '@midscene/shared/utils';
import yaml from 'js-yaml';

const debugUtils = getDebug('yaml:utils');

export function interpolateEnvVars(content: string): string {
  return content.replace(/\$\{([^}]+)\}/g, (_, envVar) => {
    const value = process.env[envVar.trim()];
    if (value === undefined) {
      throw new Error(`Environment variable "${envVar.trim()}" is not defined`);
    }
    return value;
  });
}

export function parseYamlScript(
  content: string,
  filePath?: string,
  ignoreCheckingTarget?: boolean,
): MidsceneYamlScript {
  let processedContent = content;
  if (content.indexOf('android') !== -1 && content.match(/deviceId:\s*(\d+)/)) {
    let matchedDeviceId;
    processedContent = content.replace(
      /deviceId:\s*(\d+)/g,
      (match, deviceId) => {
        matchedDeviceId = deviceId;
        return `deviceId: '${deviceId}'`;
      },
    );
    console.warn(
      `please use string-style deviceId in yaml script, for example: deviceId: "${matchedDeviceId}"`,
    );
  }
  const interpolatedContent = interpolateEnvVars(processedContent);
  const obj = yaml.load(interpolatedContent, {
    schema: yaml.JSON_SCHEMA,
  }) as MidsceneYamlScript;

  const pathTip = filePath ? `, failed to load ${filePath}` : '';
  const android =
    typeof obj.android !== 'undefined'
      ? Object.assign({}, obj.android || {})
      : undefined;
  const webConfig = obj.web || obj.target; // no need to handle null case, because web has required parameters url
  const web =
    typeof webConfig !== 'undefined'
      ? Object.assign({}, webConfig || {})
      : undefined;

  if (!ignoreCheckingTarget) {
    // make sure at least one of target/web/android is provided
    assert(
      web || android,
      `at least one of "target", "web", or "android" properties is required in yaml script${pathTip}`,
    );

    // make sure only one of target/web/android is provided
    assert(
      (web && !android) || (!web && android),
      `only one of "target", "web", or "android" properties is allowed in yaml script${pathTip}`,
    );

    // make sure the config is valid
    if (web || android) {
      assert(
        typeof web === 'object' || typeof android === 'object',
        `property "target/web/android" must be an object${pathTip}`,
      );
    }
  }

  assert(obj.tasks, `property "tasks" is required in yaml script ${pathTip}`);
  assert(
    Array.isArray(obj.tasks),
    `property "tasks" must be an array in yaml script, but got ${obj.tasks}`,
  );
  return obj;
}

export function buildDetailedLocateParam(
  locatePrompt: TUserPrompt,
  opt?: LocateOption,
): DetailedLocateParam | undefined {
  debugUtils('will call buildDetailedLocateParam', locatePrompt, opt);
  let prompt = locatePrompt || opt?.prompt || (opt as any)?.locate; // as a shortcut
  let deepThink = false;
  let cacheable = true;
  let xpath = undefined;

  if (typeof opt === 'object' && opt !== null) {
    deepThink = opt.deepThink ?? false;
    cacheable = opt.cacheable ?? true;
    xpath = opt.xpath;
    if (locatePrompt && opt.prompt && locatePrompt !== opt.prompt) {
      console.warn(
        'conflict prompt for item',
        locatePrompt,
        opt,
        'maybe you put the prompt in the wrong place',
      );
    }
    prompt = prompt || opt.prompt;
  }

  if (!prompt) {
    debugUtils(
      'no prompt, will return undefined in buildDetailedLocateParam',
      opt,
    );
    return undefined;
  }

  return {
    prompt,
    deepThink,
    cacheable,
    xpath,
  };
}

export function buildDetailedLocateParamAndRestParams(
  locatePrompt: TUserPrompt,
  opt: LocateOption | undefined,
  excludeKeys: string[] = [],
): {
  locateParam: DetailedLocateParam | undefined;
  restParams: Record<string, any>;
} {
  const locateParam = buildDetailedLocateParam(locatePrompt, opt);

  // Extract all keys from opt except the ones already included in locateParam
  const restParams: Record<string, any> = {};

  if (typeof opt === 'object' && opt !== null) {
    // Get all keys from opt
    const allKeys = Object.keys(opt);

    // Keys already included in locateParam: prompt, deepThink, cacheable, xpath
    const locateParamKeys = Object.keys(locateParam || {});

    // Extract all other keys
    for (const key of allKeys) {
      if (
        !locateParamKeys.includes(key) &&
        !excludeKeys.includes(key) &&
        key !== 'locate'
      ) {
        restParams[key] = opt[key as keyof LocateOption];
      }
    }
  }

  return {
    locateParam,
    restParams,
  };
}
