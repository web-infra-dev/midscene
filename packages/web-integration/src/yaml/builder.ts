import type {
  MidsceneYamlScript,
  MidsceneYamlScriptEnv,
  MidsceneYamlTask,
} from '@midscene/core';
import yaml from 'js-yaml';

export function buildYaml(
  env: MidsceneYamlScriptEnv,
  tasks: MidsceneYamlTask[],
) {
  const result: MidsceneYamlScript = {
    target: env,
    tasks,
  };

  return yaml.dump(result, {
    indent: 2,
  });
}
