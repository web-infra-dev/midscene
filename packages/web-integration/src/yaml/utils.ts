import assert from 'node:assert';
import yaml from 'js-yaml';

import type { MidsceneYamlScript } from '@midscene/core';
import type {
  MidsceneYamlFlowItem,
  MidsceneYamlFlowItemAIAction,
  MidsceneYamlFlowItemAIAssert,
  MidsceneYamlFlowItemAIQuery,
  MidsceneYamlFlowItemAIWaitFor,
  MidsceneYamlFlowItemSleep,
} from '@midscene/core';

function interpolateEnvVars(content: string): string {
  return content.replace(/\$\{([^}]+)\}/g, (_, envVar) => {
    const value = process.env[envVar.trim()];
    if (value === undefined) {
      throw new Error(`Environment variable "${envVar.trim()}" is not defined`);
    }
    return value;
  });
}

export function parseYamlScript(
  content: string,
  filePath?: string,
): MidsceneYamlScript {
  const interpolatedContent = interpolateEnvVars(content);
  const obj = yaml.load(interpolatedContent) as MidsceneYamlScript;
  const pathTip = filePath ? `, failed to load ${filePath}` : '';
  assert(obj.target, `property "target" is required in yaml script${pathTip}`);
  assert(
    typeof obj.target === 'object',
    `property "target" must be an object${pathTip}`,
  );
  assert(obj.tasks, `property "tasks" is required in yaml script${pathTip}`);
  assert(
    Array.isArray(obj.tasks),
    `property "tasks" must be an array${pathTip}`,
  );
  return obj;
}

export const flowItemBrief = (flowItem?: MidsceneYamlFlowItem) => {
  if (!flowItem) {
    return '';
  }

  const sliceText = (text?: string) => {
    const lengthLimit = 60;
    if (text && text.length > lengthLimit) {
      return `${text.slice(0, lengthLimit)}...`;
    }

    return text || '';
  };

  if (
    (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
    (flowItem as MidsceneYamlFlowItemAIAction).ai
  ) {
    const lastTip = (
      (flowItem as MidsceneYamlFlowItemAIAction).aiActionProgressTips || []
    ).at(-1);
    return `aiAction: ${sliceText(
      lastTip ||
        (flowItem as MidsceneYamlFlowItemAIAction).aiAction ||
        (flowItem as MidsceneYamlFlowItemAIAction).ai,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIAssert).aiAssert) {
    return `aiAssert: ${sliceText(
      (flowItem as MidsceneYamlFlowItemAIAssert).aiAssert,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery) {
    return `aiQuery: ${sliceText((flowItem as MidsceneYamlFlowItemAIQuery).aiQuery)}`;
  }
  if ((flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor) {
    return `aiWaitFor: ${sliceText(
      (flowItem as MidsceneYamlFlowItemAIWaitFor).aiWaitFor,
    )}`;
  }
  if ((flowItem as MidsceneYamlFlowItemSleep).sleep) {
    return `sleep: ${(flowItem as MidsceneYamlFlowItemSleep).sleep}`;
  }
  return '';
};
