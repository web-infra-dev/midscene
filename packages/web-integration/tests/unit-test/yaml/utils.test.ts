import { buildYaml, flowItemBrief } from '@/yaml';
import { describe, expect, test } from 'vitest';

describe('utils', () => {
  test('build yaml', () => {
    const yaml = buildYaml({ url: 'https://www.baidu.com' }, []);
    expect(yaml).toMatchSnapshot();
  });

  test('action brief text', () => {
    expect(flowItemBrief({ ai: 'search for weather' })).toMatchSnapshot();
    expect(flowItemBrief({ sleep: 1000 })).toMatchSnapshot();
    expect(
      flowItemBrief({ aiWaitFor: 'wait for something' }),
    ).toMatchSnapshot();
  });
});
