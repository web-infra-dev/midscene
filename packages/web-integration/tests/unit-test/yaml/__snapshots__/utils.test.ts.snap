// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > action brief text 1`] = `"aiAction: search for weather"`;

exports[`utils > action brief text 2`] = `"sleep: 1000"`;

exports[`utils > action brief text 3`] = `"aiWaitFor: wait for something"`;

exports[`utils > build yaml 1`] = `
"target:
  url: https://www.baidu.com
tasks: []
"
`;
